rpcs3/Emu/RSX/rsx_cache.h

```diff
@@ -494,73 +495,99 @@ namespace rsx
 			dlg->update_msg(0, getMessage(0, 0, entry_count));
 			dlg->update_msg(1, getMessage(0, 0, entry_count));
 
-			// Setup worker threads
-			unsigned nb_threads = std::thread::hardware_concurrency();
-			std::vector<std::thread> worker_threads(nb_threads);
-
 			// Preload everything needed to compile the shaders
-			// Can probably be parallelized too, but since it's mostly reading files it's probably not worth it
 			std::vector<std::tuple<pipeline_storage_type, RSXVertexProgram, RSXFragmentProgram>> unpacked;
 			std::chrono::time_point<steady_clock> last_update;
 			u32 processed_since_last_update = 0;
-
-			for (u32 i = 0; (i < entry_count) && !Emu.IsStopped(); i++)
+						
+			if (g_cfg.video.renderer == video_renderer::vulkan)
 			{
-				fs::dir_entry tmp = entries[i];
+				std::mutex inv_entries_mutex;
+				std::mutex unpacked_mutex;
+				atomic_t<u32> processed(0);
 
-				const auto filename = directory_path + "/" + tmp.name;
-				std::vector<u8> bytes;
-				fs::file f(filename);
-				if (f.size() != sizeof(pipeline_data))
-				{
-					LOG_ERROR(RSX, "Cached pipeline object %s is not binary compatible with the current shader cache", tmp.name.c_str());
-					invalid_entries.push_back(filename);
-					continue;
-				}
-				f.read<u8>(bytes, f.size());
+				// Setup worker threads
+				unsigned nb_threads = std::thread::hardware_concurrency();
+				std::vector<std::thread> worker_threads(nb_threads);
 
-				auto entry = unpack(*reinterpret_cast<pipeline_data*>(bytes.data()));
-				m_storage.preload_programs(std::get<1>(entry), std::get<2>(entry));
-				unpacked.push_back(entry);
+				std::function<void(u32)> shader_load_worker = [&](u32 index) {
+					u32 pos;
+					while (((pos = processed++) < entry_count) && !Emu.IsStopped())
+					{
+						fs::dir_entry tmp = entries[pos];
+
+						const auto filename = directory_path + "/" + tmp.name;
+						std::vector<u8> bytes;
+						fs::file f(filename);
+						if (f.size() != sizeof(pipeline_data))
+						{
+							LOG_ERROR(RSX, "Cached pipeline object %s is not binary compatible with the current shader cache", tmp.name.c_str());
+							std::lock_guard<std::mutex> lock(inv_entries_mutex);
+							invalid_entries.push_back(filename);
+							continue;
+						}
+						f.read<u8>(bytes, f.size());
+
+						auto entry = unpack(*reinterpret_cast<pipeline_data*>(bytes.data()));
+						m_storage.preload_programs(std::get<1>(entry), std::get<2>(entry));
+						{
+							std::lock_guard<std::mutex> lock(unpacked_mutex);
+							unpacked.push_back(entry);
```

Comment:
Use lock free container. See the lf_xxxx structures in utils.

Reply:
Looking into it.